pkg/plugin/util.go

```diff
@@ -0,0 +1,356 @@
+	if err != nil {
+		return parameters, "", err
+	}
+
+	// convert unstructured object into MoveEngine CR
+	var moveEngine v1alpha1.MoveEngine
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(resp.UnstructuredContent(), &moveEngine)
+	if err != nil {
+		return parameters, "", err
+	}
+
+	if moveEngine.Spec.PluginParameters != nil {
+		if err := json.Unmarshal(moveEngine.Spec.PluginParameters.Raw, &parameters); err != nil {
+			return parameters, "", fmt.Errorf("failed to Unmarshal plugin parameters")
+		}
+	} else {
+		return parameters, "", fmt.Errorf("plugin parameters not found in the MoveEngine")
+	}
+
+	return parameters, moveEngine.Spec.Mode, nil
+}
+
+func parseErrorCause(body io.ReadCloser) (*RootCause, error) {
+	var errorInfo ErrorInfo
+	data, err := ioutil.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(data, &errorInfo)
+	if err != nil {
+		return nil, err
+	}
+	return &errorInfo.Error.RootCause[0], nil
+}
+
+// insertMinioRepository patches both source and destination Elasticsearches and inject s3-repository plugin installer init-container
+func insertMinioRepository(k8sClient kubernetes.Interface, dmClient dynamic.Interface, params PluginParameters) error {
+	es, err := getElasticsearch(dmClient, params)
+	if err != nil {
+		return err
+	}
+
+	// insert plugin installer init-container
+	es.Spec.NodeSets[0].PodTemplate = corev1.PodTemplateSpec{
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name: ContainerPluginInstaller,
+					Command: []string{
+						"sh",
+						"-c",
+						"bin/elasticsearch-plugin install --batch repository-s3",
+					},
+				},
+			},
+		},
+	}
+
+	// insert minio credentials
+	es.Spec.SecureSettings = []common.SecretSource{
+		{
+			SecretName: params.Repository.Credentials,
+		},
+	}
+
+	// convert updated Elasticsearch back to unstructured object
+	updatedES, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&es)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert updated Elasticsearch into Unstructured object")
+	}
+
+	// update Elasticsearch
+	_, err = dmClient.Resource(ESGVR).Namespace(params.Elasticsearch.Namespace).Update(&unstructured.Unstructured{Object: updatedES}, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// wait for ES to be ready with the plugin installer
+	return WaitUntilElasticsearchReady(k8sClient, dmClient, params, true)
+}
+
+func WaitUntilElasticsearchReady(k8sClient kubernetes.Interface, dmClient dynamic.Interface, params PluginParameters, waitForInitContainer bool) error {
+	fmt.Println("Waiting for Elaticsearch to be ready .......")
+	err := wait.PollImmediate(5*time.Second, 20*time.Minute, func() (done bool, err error) {
+		es, err := getElasticsearch(dmClient, params)
+		if err != nil {
+			return true, err
+		}
+		if es.Status.Phase != eck.ElasticsearchReadyPhase {
+			return false, nil
+		}
+
```

Comment:
let's add a log here about waiting for elasticsearch to become ready

Reply:
done